node-db/lesson2.md

```diff
@@ -1,22 +1,390 @@
+
+## More Templating
+There is much more that can be done with **Handlebars** templating engine. Skim through the [express-handlebars](https://github.com/ericf/express-handlebars) and [handlebars documentation](handlebarsjs.com)
+
+For now, we will finish by using the **each** helper
+### Dynamically loading the templates
+
+```html
+{{#each posts}}
+    <div class="post-preview">
+        <a href="post.html">
+            <h2 class="post-title">
+                {{this.title}}
+            </h2>
+            <h3 class="post-subtitle">
+                {{this.summary}}
+            </h3>
+        </a>
+    </div>
+{{/each}}
+```
+
+in the `route`, let's load the file in `data/posts.json`
+
+```javascript
+app.get('/', function (req, res) {
+    const filePath = __dirname + '/data/posts.json';
+    const callbackFunction = function(error, file) {
+        // we call .toString() to turn the file buffer to a String
+        const fileData = file.toString();
+        // we use JSON.parse to get an object out the String
+        const postsJson = JSON.parse(fileData);
+        // send the json to the Template to render
+        res.render('index', {
+          title: 'Etzali Blog', // insert your name instead
+          posts: postsJson
+        });
+    };
+    fs.readFile(filePath, callbackFunction);
+});
+```
+
+# Callbacks
+In JavaScript, functions are first-class objects. That means they can be used in a first class manner like any other object since they are objects themselves. We can pass them as arguments to functions which is common technique in JavaScript for dealing with **asynchronous** behaviour. We have in fact already encountered callbacks when we looked at array methods such as `.forEach()`, `.map()` or `.filter`
+
+```js
+const evenNumbers = [2, 4, 6, 8];
+evenNumbers.forEach(function( num ) {
+    console.log(num);
+});
+```
+
+Here we are passing the `function( num ) {console.log(num);}` function as an argument to `.forEach()` to execute with each item in the array `evenNumbers`.
+
+The above example is **synchronous** which means that code the callback is executed immediately. Let's look at an example where the code is executed **asynchronously**. We will use a `setTimeout` function to delay execution here. In most JavaScript applications asynchronous code execution could be in response to an event such as mouse click or data coming back from a server. We will at both of those cases in detail at a future class.
+
+```js
+function delay(callback){
+    setTimeout( callback, 2000 );
+}
+
+function logRandom(){
+    const value = Math.round( Math.random() * 10);
+    console.log( value );
+    return value;
+}
+
+delay(logRandom);
+```
+
+**Key points**
+* Functions can be passed as arguments to other other function
+* Functions that are passed as arguments are called **callbacks**
+* Callbacks may be executed immediately or later
+* If callback is executed later, its return value is lost
```

Comment:
Hmm, is this last rule necessary? Why is the return value lost?

Observables are an example of a construct that asynchronously uses the return value of callbacks.